src/main/kotlin/parts/wisdom/arcraider/visualization/TaskChooser.kt

```diff
@@ -0,0 +1,87 @@
+package parts.wisdom.arcraider.visualization
+
+import java.awt.Dimension
+import java.io.File
+import javax.swing.JPanel
+import javax.swing.JScrollPane
+import javax.swing.JTree
+import javax.swing.event.TreeExpansionEvent
+import javax.swing.event.TreeExpansionListener
+import javax.swing.tree.DefaultMutableTreeNode
+import javax.swing.tree.DefaultTreeModel
+
+class FileBrowser(dir: String, onFileSelected: (File) -> Unit = { }) : JPanel() {
+
+    private lateinit var tree: JTree
+
+    private val nodeExpansionListener = object: TreeExpansionListener {
+        override fun treeExpanded(event: TreeExpansionEvent?) {
+            val subtree = (event?.path?.lastPathComponent as? DefaultMutableTreeNode) ?: return
+            if (subtree.depth < 2) {
+                for (child in subtree.children()) {
+                    TreeExplorer(child as DefaultMutableTreeNode).exploreNodes(3)
+                }
+            }
+        }
+
+        override fun treeCollapsed(event: TreeExpansionEvent?) = Unit
+    }
+
+    init {
+        val rootFile = File(dir)
+        val rootNode = DefaultMutableTreeNode(RootFileNode(rootFile))
+        tree = JTree(DefaultTreeModel(rootNode))
+        tree.showsRootHandles = true
+        tree.addTreeExpansionListener(nodeExpansionListener)
+        tree.addTreeSelectionListener {
+            val node = tree.lastSelectedPathComponent as? DefaultMutableTreeNode
+            if (node?.isLeaf == true) {
+                onFileSelected(node.file())
+            }
+        }
+
+        val scrollPane = JScrollPane(tree).apply {
+            preferredSize = Dimension(200, 500)
+        }
+        add(scrollPane)
+        TreeExplorer(rootNode).exploreNodes(3)
+    }
+
+    private inner class TreeExplorer(
```

Comment:
Why is this a class instead of just providing its methods as functions?

Reply:
I had originally copied this from a java implementation on the interwebs. I updated them to functions.